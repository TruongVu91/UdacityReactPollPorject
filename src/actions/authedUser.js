export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const LOGOUT_USER = "LOGOUT_USER";

export function setAuthedUser(authedUser) {
  return {
    type: SET_AUTHED_USER,
    authedUser,
  };
}

export function logoutUser() {
  return {
    type: LOGOUT_USER,
  };
}

export function handleLogin(username, password) {
  return (dispatch, getState) => {
    const { users } = getState();
    const user = Object.values(users).find(
      (user) => user.id === username && user.password === password
    );
    if (!!user) {
      return dispatch(setAuthedUser(user));
    }
  };
}

export function handleLogout() {
  return (dispatch) => {
    dispatch(logoutUser());
  };
}
