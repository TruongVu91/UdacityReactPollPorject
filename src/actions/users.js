export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_USER_QUESTION = "ADD_USER_QUESTION";
export const ADD_ANSWER_USER = "ADD_ANSWER_USER";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function addUserQuestion({ author, id }) {
  return {
    type: ADD_USER_QUESTION,
    author,
    id,
  };
}

export function addAnswerUser(authedUser, qid, answer) {
  return {
    type: ADD_ANSWER_USER,
    authedUser,
    qid,
    answer,
  };
}
