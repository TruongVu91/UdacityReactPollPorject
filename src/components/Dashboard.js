import { connect } from "react-redux";
import QuestionCard from "./QuestionCard";
import { Navigate } from "react-router-dom";

const Dashboard = (props) => {
  const { authedUser, questions, users, isLogin } = props;

  if (!isLogin) {
    const urlParams = new URLSearchParams(window.location.search);
    const redirectUrl = urlParams.get("redirectTo");
    return <Navigate to={redirectUrl ? redirectUrl : "/login"} />;
  }

  const newQuestions = Object.values(questions).filter((question) => {
    return (
      !question.optionOne.votes.includes(authedUser.id) &&
      !question.optionTwo.votes.includes(authedUser.id)
    );
  });

  const doneQuestions = Object.values(questions).filter((question) => {
    return (
      question.optionOne.votes.includes(authedUser.id) ||
      question.optionTwo.votes.includes(authedUser.id)
    );
  });

  console.log("newQuestions", questions);

  return (
    <div>
      <h3 className="h-body">New Questions</h3>
      <div className="card-holder">
        {newQuestions.map((question) => {
          return (
            <QuestionCard
              key={question.id}
              question={question}
              author={users[question.author]}
            />
          );
        })}
      </div>
      <h3 className="h-body">Done</h3>
      <div className="card-holder">
        {doneQuestions.map((question) => {
          return (
            <QuestionCard
              key={question.id}
              question={question}
              author={users[question.author]}
            />
          );
        })}
      </div>
    </div>
  );
};

const mapStateToProps = ({ authedUser, questions, users }) => ({
  authedUser,
  questions: Object.values(questions).sort((a, b) => b.timestamp - a.timestamp),
  users,
  isLogin: !!authedUser,
});

export default connect(mapStateToProps)(Dashboard);
