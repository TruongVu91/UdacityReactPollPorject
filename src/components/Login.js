import { useState } from "react";
import { connect } from "react-redux";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import { handleLogin } from "../actions/authedUser";
import { Navigate, useLocation, useNavigate } from "react-router-dom";

const Login = ({ dispatch, isLogin, users }) => {
  const [userId, setUserId] = useState("");
  const [password, setPassword] = useState("");
  const navigate = useNavigate();
  const { state } = useLocation();

  if (isLogin) {
    const urlParams = new URLSearchParams(window.location.search);
    const redirectUrl = urlParams.get("redirectTo");
    return <Navigate to={redirectUrl ? redirectUrl : "/"} />;
  }

  const handelSelectChange = (e) => {
    const text = e.target.value;
    const chosenOne = users.find((user) => user.name === text);
    setUserId(chosenOne.id);
    setPassword(chosenOne.password);
  };

  const handleLoginUser = (e) => {
    e.preventDefault();
    dispatch(handleLogin(userId, password));
    setUserId("");
    setPassword("");
    navigate(state?.path || "/");
  };

  return (
    <div>
      <h2 className="h-body" data-testid="loginTitle">
        Login
      </h2>
      <Form onSubmit={handleLoginUser}>
        <Form.Label>Select User</Form.Label>
        <Form.Select
          className="mb-3"
          onChange={handelSelectChange}
          data-testid="userSelect"
        >
          <option>Chose user</option>
          {users.map((user) => {
            return (
              <option key={user.id} value={user.name}>
                {user.name}
              </option>
            );
          })}
        </Form.Select>
        <Form.Group className="mb-3">
          <Form.Label>User ID</Form.Label>
          <Form.Control
            type="input"
            placeholder="Enter user ID"
            value={userId}
            disabled
            data-testid="userid"
          />
        </Form.Group>
        <Form.Group className="mb-3">
          <Form.Label>Password</Form.Label>
          <Form.Control
            type="input"
            placeholder="Enter password"
            value={password}
            disabled
            data-testid="password"
          />
        </Form.Group>
        <Button
          variant="primary"
          type="submit"
          disabled={userId === "" || password === ""}
          data-testid="loginButton"
        >
          Submit
        </Button>
      </Form>
    </div>
  );
};

const mapStateToProps = ({ users, authedUser }) => {
  console.log("users test", users);
  return {
    isLogin: !!authedUser,
    users: Object.values(users),
  };
};

export default connect(mapStateToProps)(Login);
