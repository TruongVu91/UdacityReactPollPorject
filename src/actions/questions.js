import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { addAnswerUser, addUserQuestion } from "./users";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

function addAnswerQuestion(author, qid, answer) {
  return {
    type: ADD_ANSWER_QUESTION,
    author,
    qid,
    answer,
  };
}

export function handleAddQuestion(optionOneText, optionTwoText) {
  console.log("optionOneText", optionOneText);
  console.log("optionTwoText", optionTwoText);
  return (dispatch, getState) => {
    const { authedUser } = getState();
    return saveQuestion({
      optionOneText,
      optionTwoText,
      author: authedUser,
    }).then((question) => {
      dispatch(addQuestion(question));
      dispatch(addUserQuestion(question));
    });
  };
}

export function handleAddAnswer(questionId, answer) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    return saveQuestionAnswer({
      authedUser: authedUser.id,
      qid: questionId,
      answer,
    }).then(() => {
      dispatch(addAnswerQuestion(authedUser.id, questionId, answer));
      dispatch(addAnswerUser(authedUser.id, questionId, answer));
    });
  };
}
