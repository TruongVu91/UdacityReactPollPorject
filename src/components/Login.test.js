/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/render-result-naming-convention */
import "@testing-library/jest-dom/extend-expect";

import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import middleware from "../middleware";
import reducers from "../reducers";
import Login from "./Login";
import { BrowserRouter } from "react-router-dom";
import { handleInitialData } from "../actions/shared";
import Dashboard from "./Dashboard";

const store = createStore(reducers, middleware);

describe("Login", () => {
  it("check login screen display", () => {
    const loginView = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    expect(loginView).toBeDefined();
    expect(loginView).toMatchSnapshot();
  });

  it("check login element display", async () => {
    await store.dispatch(handleInitialData());
    const loginView = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    const loginTitleElement = loginView.getByTestId("loginTitle");
    const usernIdInputElement = loginView.getByTestId("userid");
    const passwordInputElement = loginView.getByTestId("password");
    const loginButtonElement = loginView.getByTestId("loginButton");
    expect(loginTitleElement).toBeInTheDocument();
    expect(usernIdInputElement).toBeInTheDocument();
    expect(passwordInputElement).toBeInTheDocument();
    expect(loginButtonElement).toBeInTheDocument();
  });

  it("check login page element dissapear after click login", async () => {
    await store.dispatch(handleInitialData());

    const loginView = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    const loginButtonElement = loginView.getByTestId("loginButton");
    const userSelectElement = loginView.getByTestId("userSelect");

    fireEvent.change(userSelectElement, { target: { value: "Sarah Edo" } });

    fireEvent.click(loginButtonElement);

    expect(userSelectElement).not.toBeInTheDocument();
  });
});
