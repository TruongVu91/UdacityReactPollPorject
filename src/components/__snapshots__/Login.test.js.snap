// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login check login screen display 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h2
          class="h-body"
          data-testid="loginTitle"
        >
          Login
        </h2>
        <form
          class=""
        >
          <label
            class="form-label"
          >
            Select User
          </label>
          <select
            class="mb-3 form-select"
            data-testid="userSelect"
          >
            <option>
              Chose user
            </option>
          </select>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              User ID
            </label>
            <input
              class="form-control"
              data-testid="userid"
              disabled=""
              placeholder="Enter user ID"
              type="input"
              value=""
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              Password
            </label>
            <input
              class="form-control"
              data-testid="password"
              disabled=""
              placeholder="Enter password"
              type="input"
              value=""
            />
          </div>
          <button
            class="btn btn-primary"
            data-testid="loginButton"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h2
        class="h-body"
        data-testid="loginTitle"
      >
        Login
      </h2>
      <form
        class=""
      >
        <label
          class="form-label"
        >
          Select User
        </label>
        <select
          class="mb-3 form-select"
          data-testid="userSelect"
        >
          <option>
            Chose user
          </option>
        </select>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
          >
            User ID
          </label>
          <input
            class="form-control"
            data-testid="userid"
            disabled=""
            placeholder="Enter user ID"
            type="input"
            value=""
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
          >
            Password
          </label>
          <input
            class="form-control"
            data-testid="password"
            disabled=""
            placeholder="Enter password"
            type="input"
            value=""
          />
        </div>
        <button
          class="btn btn-primary"
          data-testid="loginButton"
          disabled=""
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
